### DISCUSSION TITLE
How to fill outside of a polygon?

### DISCUSSION THREAD

**Discussion by talhazengin** [SixLabors/ImageSharp.Drawing#63] [Q&A] [answered]
I could not do it.
Help me thanks.

**Comment by talhazengin** [comment]
I did this, but this methodology does not perform well. I need new ideas.

            for (int y = 0; y < image.Height; y++)
            {
                Span<Rgba32> pixelRowSpan = image.GetPixelRowSpan(y);
                
                for (int x = 0; x < image.Width; x++)
                {
                    if (!polygon.Contains(new PointF(x, y)))
                    {
                        pixelRowSpan[x] = myColorRgba;
                    }
                }
            }

**Reply by tocsoft** [reply]
yeah don't do that, you can just clip the path out of a rectangle the size of your canvas... i've an example in my answer

**Comment by tocsoft** [comment, marked as answer]
The process would be to create a rectangle the size of your canvas, then `clip` from that rectangle the polygon you have this will give you a new polygon of the negative space, you then just fill it. this is an example extension method that can do it for you for simple colors.

```csharp
public static class ImageExtensions
{
    public static IImageProcessingContext FillOuter(
        this IImageProcessingContext source,
        Color brush,
        IPath path)
    {
        var rectangle = new RectangularPolygon(PointF.Empty, source.GetCurrentSize());
        var clipped = rectangle.Clip(path);
        return source.Fill(brush, clipped);
    }
}
```

**An example of use**

_Black circle on white background_
![InnerFill2](https://user-images.githubusercontent.com/166440/86224533-e62ae400-bb80-11ea-9345-de2f89b2decc.png)
```csharp
var circle1 = new EllipsePolygon(new PointF(100, 100), 50);
var circle3 = new EllipsePolygon(new PointF(100, 100), 20);

var path2 = circle1.Clip(circle3);
using (var img = new Image<Rgba32>(300, 300, Color.White))
{
    img.Mutate(x => x.Fill(Color.Black, path2));
    img.Save("InnerFill2.png"); 
}
```

_and the inverse fill_
![InnerFillOuter2](https://user-images.githubusercontent.com/166440/86224532-e5924d80-bb80-11ea-8df7-bf518ebd942f.png)
```csharp
var circle1 = new EllipsePolygon(new PointF(100, 100), 50);
var circle3 = new EllipsePolygon(new PointF(100, 100), 20);

var path2 = circle1.Clip(circle3);
using (var img = new Image<Rgba32>(300, 300, Color.White))
{
    img.Mutate(x => x.FillOuter(Color.Black, path2));
    img.Save("InnerFillOuter2.png");
}

```

**Reply by talhazengin** [reply]
Thank you, good answer.
